### PR TITLE
Article adding feature implemented

### PR DESCRIPTION
**Pull request by pavel-temtsunik**

### CODE REVIEW

File: features/step_definitions/article_steps.rb [hunk 1]
```diff
@@ -6,10 +6,40 @@
...
 
+#############################################################
+#                      ACTIONS                              #
+#############################################################
+
+When /^I click new article button on articles list page$/ do
+  ArticlesListPage.given.add_new_article
+end
+
+And /^I fill new article form on new article page with data:$/ do |table|
```
Comment (romikoops) [inline features/step_definitions/article_steps.rb]: I want to see:
And /^I fill form on new article page with data:$/ do |table|

File: features/step_definitions/article_steps.rb [hunk 2]
```diff
@@ -6,10 +6,40 @@
...
+ article = table.rows_hash.symbolize_keys
+  NewArticlePage.given.fill_form(article[:title], article[:text])
+
```
Comment (romikoops) [inline features/step_definitions/article_steps.rb]: remove empty line

File: features/step_definitions/article_steps.rb [hunk 3]
```diff
@@ -6,10 +6,40 @@
...
+end
+
+And /^I submit new article form on new article page$/ do
```
Comment (romikoops) [inline features/step_definitions/article_steps.rb]: When /^I submit form on new article page$/ do

File: features/step_definitions/article_steps.rb [hunk 4]
```diff
@@ -6,10 +6,40 @@
...
-end
\ No newline at end of file
+end
+
+Then /^I should see article parameters on article page with data:$/ do |table|
+  article = table.rows_hash.symbolize_keys
+ (expect(ArticlePage.given.text).to include(article[:title]) and expect(ArticlePage.given.text).to include(article[:text]))
+end
+
+Then(/^I should see article on articles list page with data:$/) do |table|
```
Comment (romikoops) [inline features/step_definitions/article_steps.rb]: Remove () here

File: features/step_definitions/article_steps.rb [hunk 5]
```diff
@@ -6,10 +6,40 @@
...
 #              CHECKS              #
 ####################################
 
 Then /^I see comment displayed on (.*) page:$/ do |page, table|
   expect(page.given.comment_data).to eql(table.rows_hash.symbolize_keys)
-end
\ No newline at end of file
+end
+
+Then /^I should see article parameters on article page with data:$/ do |table|
```
Comment (romikoops) [inline features/step_definitions/article_steps.rb]: Then /^I should see article on article page with data:$/ do |table|

File: features/step_definitions/article_steps.rb [hunk 6]
```diff
@@ -6,10 +6,40 @@
...
+  article = table.rows_hash.symbolize_keys
+ (expect(ArticlePage.given.text).to include(article[:title]) and expect(ArticlePage.given.text).to include(article[:text]))
+end
+
+Then(/^I should see article on articles list page with data:$/) do |table|
+  article = table.rows_hash.symbolize_keys
+  ArticlesListPage.open
```
Comment (romikoops) [inline features/step_definitions/article_steps.rb]: WTF? Action+Check in Then??? 
It should be When + Then

File: features/support/env.rb [hunk 7]
```diff
@@ -1,9 +1,11 @@
 require 'cucumber'
 require 'capybara/cucumber'
 require_relative '../../boot'
+require_relative "gen"
 
 World(Capybara::Settings)
 World(DataGenerator)
+include Gen
```
Comment (romikoops) [inline features/support/env.rb]: Bad idea, remove it

File: features/support/gen.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
+module Gen
```
Comment (romikoops) [inline features/support/gen.rb]: Use the same structure as for user generator

module DataGenerator
  module Gen
     class << self

File: features/support/gen.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  def gen_article_title(serial=nil)
+    gen_entity('article_title', serial)
+  end
+
+  def gen_article_text(serial=nil)
+    gen_entity('article_text', serial*7)
+  end
+
+  class Article < Object
```
Comment (romikoops) [inline features/support/gen.rb]: What for Object here? It is default class for any new class

File: features/support/transformers.rb [hunk 10]
```diff
@@ -1,12 +1,17 @@
...
-  res
+def define_uniq_entity_transform name
+  Transform /UNIQ_#{name}(\d*)(?:\[\:(.+)\])?/ do |num, property|
+    res = Gen.send("given_#{name.downcase}_by_number", num)
+    res = res.send(property) if property
+    res
+  end
 end
 
+define_uniq_entity_transform 'ARTICLE'
```
Comment (romikoops) [inline features/support/transformers.rb]: WTF? ))))) Remove it

File: features/support/transformers.rb [hunk 11]
```diff
@@ -1,12 +1,17 @@
...
+define_uniq_entity_transform 'USER'
```
Comment (romikoops) [inline features/support/transformers.rb]: +1

File: pages/articles_list_page.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
+class ArticlesListPage < WebPage
+  URL = '/articles/'
+  validates :title, pattern: /\ADemo web application - Listing Articles\z/
+  validates :url, pattern: /\/articles/
```
Comment (romikoops) [inline pages/articles_list_page.rb]: Why 2 validators here? remove url

File: features/support/gen.rb (hunk 9 above)
Comment (pavel-temtsunik) [inline features/support/gen.rb]: http://prntscr.com/6wnl4m this is a  screenshot from original module  \lib\ruby\gems\2.1.0\gems\howitzer-1.0.2\lib\howitzer\utils\data_generator\gen.rb
I've done the same. Isn't that right?

File: features/support/env.rb (hunk 7 above)
Comment (pavel-temtsunik) [inline features/support/env.rb]: Then i've deleted "include Gen"  , i get such error:
"NoMethodError: undefined method `given_article_by_number' for DataGenerator::Gen:Module"
Maybe i need replace it with another string?

File: features/support/transformers.rb (hunk 10 above)
Comment (pavel-temtsunik) [inline features/support/transformers.rb]: 08.04.2015 you've sent me example of "transformers.rb" via skype.
I've taken this code snippet from that example file. http://prntscr.com/6wnpge  Isn't that right?

File: features/step_definitions/article_steps.rb (hunk 1 above)
Comment (pavel-temtsunik) [inline features/step_definitions/article_steps.rb]: +

File: features/step_definitions/article_steps.rb (hunk 2 above)
Comment (pavel-temtsunik) [inline features/step_definitions/article_steps.rb]: ready

File: features/step_definitions/article_steps.rb (hunk 3 above)
Comment (pavel-temtsunik) [inline features/step_definitions/article_steps.rb]: +

File: features/step_definitions/article_steps.rb (hunk 4 above)
Comment (pavel-temtsunik) [inline features/step_definitions/article_steps.rb]: +

File: features/step_definitions/article_steps.rb (hunk 5 above)
Comment (pavel-temtsunik) [inline features/step_definitions/article_steps.rb]: +

File: features/step_definitions/article_steps.rb (hunk 6 above)
Comment (pavel-temtsunik) [inline features/step_definitions/article_steps.rb]: +

File: pages/articles_list_page.rb (hunk 12 above)
Comment (pavel-temtsunik) [inline pages/articles_list_page.rb]: +

File: features/support/env.rb (hunk 7 above)
Comment (romikoops) [inline features/support/env.rb]: I looked old code and found it should has such structure
'''ruby
module Gen
  class << self
'''

File: features/support/gen.rb [hunk 13]
```diff
@@ -0,0 +1,41 @@
+module DataGenerator
```
Comment (romikoops) [inline features/support/gen.rb]: remove this line

File: features/support/gen.rb (hunk 9 above)
Comment (romikoops) [inline features/support/gen.rb]: I guess it was workaround, anyway the functionality will be removed soon

File: features/support/transformers.rb (hunk 10 above)
Comment (romikoops) [inline features/support/transformers.rb]: or, I realized what for it was created. So, keep it as it is

File: features/support/env.rb (hunk 7 above)
Comment (pavel-temtsunik) [inline features/support/env.rb]: +

File: features/support/gen.rb (hunk 13 above)
Comment (pavel-temtsunik) [inline features/support/gen.rb]: +

File: features/support/gen.rb (hunk 9 above)
Comment (pavel-temtsunik) [inline features/support/gen.rb]: +

File: features/support/transformers.rb (hunk 10 above)
Comment (pavel-temtsunik) [inline features/support/transformers.rb]: +

File: features/step_definitions/article_steps.rb [hunk 14]
```diff
@@ -12,4 +19,10 @@
 
 Then /^I see comment displayed on (.*) page:$/ do |page, table|
   expect(page.given.comment_data).to eql(table.rows_hash.symbolize_keys)
-end
\ No newline at end of file
+end
+
+Then /^I should see article on (.+) page with data:$/ do |page, table|
+  article = table.rows_hash.symbolize_keys
+ expect(page.given.text).to include(article[:title]) and expect(page.given.text).to include(article[:text])
```
Comment (romikoops) [inline features/step_definitions/article_steps.rb]: What for is and there? Remove it

File: features/support/env.rb [hunk 15]
```diff
@@ -1,6 +1,7 @@
 require 'cucumber'
 require 'capybara/cucumber'
 require_relative '../../boot'
+require_relative "gen"
```
Comment (romikoops) [inline features/support/env.rb]: All files from /support are loaded automatically, remove it

File: features/support/gen.rb [hunk 16]
```diff
@@ -0,0 +1,39 @@
+  module Gen
+    class << self
+
+  def article(params={})
```
Comment (romikoops) [inline features/support/gen.rb]: Why is it shifted from left side?

File: pages/articles_list_page.rb [hunk 17]
```diff
@@ -0,0 +1,17 @@
+class ArticlesListPage < WebPage
+  URL = '/articles/'
+  validates :title, pattern: /\ADemo web application - Listing Articles\z/
+
+
```
Comment (romikoops) [inline pages/articles_list_page.rb]: Why so big amount of empty lines?

File: pages/articles_list_page.rb [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+  add_locator :new_article_button,  xpath: "//a[@href='/articles/new']"
+
+
+  def add_new_article
+    log.info "Adding new article"
+    find(locator(:new_article_button)).click
+    NewArticlePage.given
+  end
+
+
```
Comment (romikoops) [inline pages/articles_list_page.rb]: the same

File: pages/articles_list_page.rb [hunk 19]
```diff
@@ -0,0 +1,17 @@
+class ArticlesListPage < WebPage
+  URL = '/articles/'
+  validates :title, pattern: /\ADemo web application - Listing Articles\z/
+
+
+
+  add_locator :new_article_button,  xpath: "//a[@href='/articles/new']"
+
+
```
Comment (romikoops) [inline pages/articles_list_page.rb]: the same

File: features/support/gen.rb [hunk 20]
```diff
@@ -0,0 +1,39 @@
+module Gen
+  class << self
+
+  def article(params={})
```
Comment (romikoops) [inline features/support/gen.rb]: + 2 spaces for all methods

**Pull request merged**